samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,751 @@
+	return ctx->sock;
+
+error:
+	if (ctx->sock >= 0) {
+		zsock_close(ctx->sock);
+		ctx->sock = -1;
+	}
+	zsock_freeaddrinfo(res);
+	return -errno;
+}
+
+/*
+ * @brief Callback for HTTP response processing
+ *
+ * Processes received HTTP response data in chunks and stores them for firmware update.
+ *
+ * @param rsp Response structure containing received data
+ * @param final_data Flag indicating if this is the final fragment
+ * @param user_data User context data (pointer to app_ctx)
+ * @return 0 on success, negative error code on failure
+ */
+
+static int ota_http_response_cb(struct http_response *rsp, enum http_final_call final_data,
+				void *user_data)
+{
+	struct app_ctx *ctx = (struct app_ctx *)user_data;
+
+	/* Check for HTTP errors */
+	if (rsp->http_status_code >= 400) {
+		printf("HTTP error: %d %s\n", rsp->http_status_code, rsp->http_status);
+		ctx->http_status_code = rsp->http_status_code;
+		return -ENODATA;
+	}
+
+	if (rsp->http_status_code == 301) {
+		printf("HTTP redirect: %d %s\n", rsp->http_status_code, rsp->http_status);
+		ctx->http_status_code = rsp->http_status_code;
+		return -EPROTO;
+	}
+
+	ctx->ota_image_size = rsp->content_range.total;
+
+	if (rsp->body_frag_len > 0) {
+		/* Check if we have space in flash_buffer */
+		if (ctx->flash_buffer_len + rsp->body_frag_len > sizeof(ctx->flash_buffer)) {
+			return -ENODATA;
+		}
+		memcpy(ctx->flash_buffer + ctx->flash_buffer_len, rsp->body_frag_start,
+		       rsp->body_frag_len);
+		ctx->flash_buffer_len += rsp->body_frag_len;
+	}
+
+	return 0;
+}
+/*
+ * @brief Download firmware chunk from server
+ *
+ * @param ctx Pointer to application context
+ * @return int 0 on success, negative error code on failure
+ */
+static int ota_download_firmware_chunk(struct app_ctx *ctx)
+{
+	char range_header[64];
+	const char *headers[2] = {range_header, NULL};
+	int ret = 0;
+	struct http_request req = {
+		.method = HTTP_GET,
+		.url = ctx->http_parse_info.path,
+		.header_fields = headers,
+		.host = ctx->http_parse_info.host,
+		.protocol = "HTTP/1.1",
+		.response = ota_http_response_cb,
+		.recv_buf = ctx->http_recv_buffer,
+		.recv_buf_len = 1024
+	};
+
+	/* Reset flash buffer length for new chunk */
+	ctx->flash_buffer_len = 0;
+
+	/* Prepare headers */
+	snprintf(range_header, sizeof(range_header), "Range: bytes=%d-%d\r\n",
+		 ctx->ota_range_start_byte, ctx->ota_range_end_byte);
+
+	/* Send request */
+	ret = http_client_req(ctx->sock, &req, 30000, ctx);
+	if (ret < 0) {
+		printf("HTTP request failed: %d (%s), requested range=%d-%d\n",
+		       ctx->http_status_code,
+		       (ctx->http_status_code >= 100) ? "HTTP error" : strerror(-ret),
+		       ctx->ota_range_start_byte, ctx->ota_range_end_byte);
+	}
+
+	if (ctx->http_status_code >= 400) {
+		ret = -ENODATA;
+	}
+	if (ctx->http_status_code == 301) {
+		ret = -EPROTO;
+	}
+	return ret;
+}
+
+static bool ota_update_http_range_header(struct app_ctx *ctx)
+{
+	ctx->ota_range_start_byte = ctx->ota_range_end_byte + 1;
```

Comment:
This is probably unnecessary twisted. `ctx->ota_range_start_byte += sizeof(ctx->flash_buffer);` is easier to understand (sizeof(...) is a constant data while `ctx->ota_range_end_byte` may have been updated somewhere in the algorithm)